rasa/core/trackers.py

```diff
@@ -383,31 +384,53 @@ def _undo_till_previous(event_type: Type[Event], done_events: List[Event]) -> No
             if isinstance(e, event_type):
                 break
 
-    @staticmethod
-    def _first_loop_execution_or_previous_rejection(
-        loop_action_name: Text, applied_events: List[Event]
+    def _first_loop_execution_or_unhappy_path(
+        self, loop_action_name: Text, applied_events: List[Event]
     ) -> bool:
+        next_action: Optional[Text] = None
+
         for event in reversed(applied_events):
-            # Form was rejected previously.
-            if (
-                isinstance(event, ActionExecutionRejected)
-                and event.action_name == loop_action_name
-            ):
+            # If there was a form deactivation we know that the form is running for
+            # the first time.
```

Comment:
Rather smth along the lines:
We should look for previous form actions only until previous form deactivation event, because it means that further there was a different form.